packages/grafana-ui/src/components/Modal/Modal.tsx

```diff
@@ -19,48 +19,51 @@ export interface Props extends Themeable {
 
   /** If not set will call onDismiss if that is set. */
   onClickBackdrop?: () => void;
+  render?: (props: RenderApi) => JSX.Element;
 }
 
-export class UnthemedModal extends React.PureComponent<Props> {
-  onDismiss = () => {
-    if (this.props.onDismiss) {
-      this.props.onDismiss();
-    }
-  };
-
-  onClickBackdrop = () => {
-    this.onDismiss();
-  };
+interface RenderApi {
+  styles: ReturnType<typeof getModalStyles>;
+}
 
-  renderDefaultHeader(title: string) {
-    const { icon, iconTooltip } = this.props;
+export function Modal(props: PropsWithChildren<Props>): JSX.Element | null {
+  const { title, children, isOpen = false, className, onDismiss: propsOnDismiss, onClickBackdrop, render } = props;
+  const theme = useTheme();
+  const styles = getModalStyles(theme);
+  const onDismiss = useCallback(() => {
+    if (propsOnDismiss) {
+      propsOnDismiss();
+    }
+  }, [propsOnDismiss]);
 
-    return <ModalHeader icon={icon} iconTooltip={iconTooltip} title={title} />;
+  if (!isOpen) {
+    return null;
   }
 
-  render() {
-    const { title, isOpen = false, theme, className } = this.props;
-    const styles = getModalStyles(theme);
-
-    if (!isOpen) {
-      return null;
-    }
-
-    return (
-      <Portal>
-        <div className={cx(styles.modal, className)}>
-          <div className={styles.modalHeader}>
-            {typeof title === 'string' ? this.renderDefaultHeader(title) : title}
-            <div className={styles.modalHeaderClose}>
-              <IconButton surface="header" name="times" size="lg" onClick={this.onDismiss} />
-            </div>
+  return (
+    <Portal>
+      <div className={cx(styles.modal, className)}>
+        <div className={styles.modalHeader}>
+          {typeof title === 'string' && <DefaultModalHeader {...props} title={title} />}
+          {typeof title !== 'string' && title}
+          <div className={styles.modalHeaderClose}>
+            <IconButton surface="header" name="times" size="lg" onClick={onDismiss} />
           </div>
-          <div className={styles.modalContent}>{this.props.children}</div>
         </div>
-        <div className={styles.modalBackdrop} onClick={this.props.onClickBackdrop || this.onClickBackdrop} />
-      </Portal>
-    );
-  }
+        {render && render({ styles })}
+        {!render && <div className={styles.modalContent}>{children}</div>}
```

Comment:
I think this might get a bit confusing when users provide both render and children prop. Maybe adding a warning, if both props are present, that says that the render prop is used, would be a good idea here?

Reply:
Great suggestion, but this depends on if I'll go ahead and refactor all existing Modals or not. But I guess that based on your feedback we should either refactor all Modal or introduce yet another className prop. The latter one is something that we should avoid IMHO.

Reply:
@dprokop would love your feedback on this before I take my next steps

Reply:
Agree with Alex that this solution adds some unnecessary complexity to the Moda. the fundamental problem is really that the Select does not support portal for the dropdown.

I would add an optional property on the Modal component to disable the overflow: auto property in the modal content, or add a contentClassName, not sure which is best out of those two :)